### PR TITLE
Successful implementation of data processing assignment

### PR DESCRIPTION
**Pull request by nejidevelops**
### Create models
-  Create model classes for all entities as shown in the [ERD diagram](../images/blog_app_erd_v1_1.png).
-  Set up associations between models.
    - Remember that `author_id` column in `posts` table should be the foreign_key for the `users` table.

### Use models to insert data
- Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
- Open rails console for your project.
- Create at least one user by running the following code:
    ```ruby
       first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
       second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')
    ```
- Create at least 4 posts written by one of the users you created by running the following code:
    ```ruby
       first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
    ```
- Create at least 6 posts comments for one of the posts you created by running the following code:
    ```ruby
       Comment.create(post: first_post, user: second_user, text: 'Hi Tom!' )
    ```
- Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

### Create custom methods
Your models should include some additional methods.
-  Users
    - A method that returns the 3 most recent posts for a given user.
-  Posts
    - A method that updates the posts counter for a user.
    - A method which returns the 5 most recent comments for a given post.
-  Comments
    - A method that updates the comments counter for a post.
-  Likes
    - A method that updates the likes counter for a post.
- Go to `rails c` and check if your methods are working.

```sh
rails generate model User name:string photo:string bio:text posts_counter:integer
rails generate model Post author:references title:string text:text comments_counter:integer likes_counter:integer
rails generate model Comment user:references post:references text:string
rails generate model Like user:references post:references
```

### app/models/user.rb
```sh
class User < ApplicationRecord
  has_many :posts, foreign_key: 'author_id'
  has_many :comments
  has_many :likes
end
```

### app/models/post.rb
```sh
class Post < ApplicationRecord
  belongs_to :author, class_name: 'User'
  has_many :comments
  has_many :likes
end
```

### app/models/comment.rb
```sh
class Comment < ApplicationRecord
  belongs_to :user
  belongs_to :post
end
```

### app/models/like.rb
```sh
class Like < ApplicationRecord
  belongs_to :user
  belongs_to :post
end
```

- Insert Data Using the Rails Console:
- Run rails c to open the Rails console, and then execute the following commands to create users, posts, and comments:

```sh
first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')

first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')

Comment.create(post: first_post, user: second_user, text: 'Hi Tom!')
```

- Implement Custom Methods in Model Classes:
- Add the custom methods to your model classes as specified in your requirements:

# app/models/user.rb
```sh
class User < ApplicationRecord
  # ... (previous associations)

  def three_most_recent_posts
    posts.order(created_at: :desc).limit(3)
  end
end
```

# app/models/post.rb
```sh
class Post < ApplicationRecord
  # ... (previous associations)

  def update_posts_counter
    author.update(posts_counter: author.posts.count)
  end

  def five_most_recent_comments
    comments.order(created_at: :desc).limit(5)
  end
end
```

# app/models/comment.rb
```sh
class Comment < ApplicationRecord
  # ... (previous associations)
end
```

### Verify the Custom Methods:
- In the Rails console, you can verify that your custom methods work as expected:

# Check the three most recent posts for a user (replace 1 with the user's ID)
```sh
User.find(1).three_most_recent_posts
```

# Update the posts counter for a user (replace 1 with the user's ID)
```sh
Post.find(1).update_posts_counter
```

# Check the five most recent comments for a post (replace 1 with the post's ID)
```sh
Post.find(1).five_most_recent_comments
```

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+end
```
Comment (uwadonat) [inline app/models/comment.rb]: - [ ] From the project requirement, you should have `update_comments_counter` method and make it private.

def update_comments_counter
  post.increment!(:comments_counter)
end

- [ ] Also I recommend you use `Rails Active Callbacks`- `after_save` to call that update method every time a new comment is saved to database.

 class Comment < ApplicationRecord
   ` after_save : update_comments_counter`
....
    private
     def update_comments_counter
      ...
    end
end

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+
+  validates :user_id, presence: true
+  validates :post_id, presence: true
+end
```
Comment (uwadonat) [inline app/models/like.rb]: - [ ] Let's add `update_likes_counter`. Make it private and use `after_save` callback as described above.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comments
+  has_many :likes
+
+  def update_posts_counter
+    author.update(posts_counter: author.posts.count)
+  end
```
Comment (uwadonat) [inline app/models/post.rb:12]: - [ ] `update_posts_counter` should be private. also use `after_save` callback as described above.

**Review by uwadonat — CHANGES_REQUESTED** [review]
## Hi @mohammadnewton,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
- Your commit messages are very descriptive ✔️
- Clean and well-organized project ✔️
- Correct workflow was used ✔️
- No linter errors ✔️
almost

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ShadyShawkat — APPROVED** [review]
## Hi @mohammadnewton, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 
- Your readme looks professional ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**